src/Umbraco.Infrastructure/Persistence/Repositories/Implement/ContentRepositoryBase.cs

```diff
@@ -767,8 +767,20 @@ protected string GetQuotedFieldName(string tableName, string fieldName)
 
         #region UnitOfWork Events
 
-        // TODO: The reason these events are in the repository is for legacy, the events should exist at the service
-        // level now since we can fire these events within the transaction... so move the events to service level
+        /*
+         * TODO: The reason these events are in the repository is for legacy, the events should exist at the service
+         * level now since we can fire these events within the transaction...
+         * The reason these events 'need' to fire in the transaction is to ensure data consistency with Nucache (currently
+         * the only thing that uses them). For example, if the transaction succeeds and NuCache listened to ContentService.Saved
+         * and then NuCache failed at persisting data after the trans completed, then NuCache would be out of sync. This way
```

Comment:
Why would NuCache fail to persist data?

I may be reading this comment  wrong but I don't think a failure of an event subscriber should trigger a rollback of a persistence layer,  operation; it's entirely back-to-front. Caches and the content queries that use them should be designed to handle misses. 

Save should always happen.

Reply:
This is what we were talking about on the events chat and whether events should also be raised within the trans or not. NuCache manages data in the database along with it's in-memory/disk caches. The idea here is that the DB for nucache should never be inconsistent with what is in the DB for content so a rollback will trigger if it fails because these events are raised within the trans.

I agree with what you are saying and that is what I would have done, but based on discussions with various folks a while back the 'safe' thing to do is to have any DB changes related to the current events done within the transaction so everything is guaranteed consistent. This is partly what @lars-erik was referring to in the events chat https://github.com/umbraco/Umbraco-CMS/issues/9397#issuecomment-741677912 

But the thing is, that only guarantees consistency at the DB level, there are other non-DB things that shouldn't run in the transaction like how lucene indexes get populated. 

I wrote all these code comments here because I had to re-investigate all this and how/why it was done so at least we have a clue why they exist.

Reply:
>I agree with what you are saying and that is what I would have done, but based on discussions with various folks a while back the 'safe' thing to do is to have any DB changes related to the current events done within the transaction so everything is guaranteed consistent.

That's most unfortunate. I would definitely aim to refactor to remove transactional coupling in this manner. Repositories should certainly raise an event containing a collection of entities updated during and after a transaction but NuCache et all, should really only be awaiting notifications for once a transaction is complete. If something fails they can log. A `PublishedContentService` can then handle cache misses.

Reply:
Agreed. In a load balanced scenario nucache is going to be out of sync for some amount of time anyway.

Reply:
> In a load balanced scenario nucache is going to be out of sync for some amount of time anyway.

This isn't about nucache's in-memory caches though, this is about nucache's source data which is in the database. Having it within the transaction guarantees that the nucache source data is always consistent with the umbraco source data.

Would like to hear @lars-erik 's opinion here since I believe he is for having some events such as this to be raised within the trans for this exact purpose. I see he's added 🚀 to james' comment but pretty sure that is different to what was mentioned here https://github.com/umbraco/Umbraco-CMS/issues/9397#issuecomment-741677912?

Reply:
Is there a problem with eventual consistency here?

Reply:
I believe Umbraco have always benefited from treating content data and content cache as separate bounded contexts if you will. Which means, if we wanna follow good practices, that there should only be serializable messages passed between them.

Events that mutate the _entities_ could/should be ran in the same transaction. Events that perform operations in other bounded context should be possible to run OOP and in separate transactions. IE. write NuCache source data.  
NuCache is already toppling over, going out of sync etc. making us click those red buttons from bottom to top. The existing "validity layer" dependent of in-transaction and in-proc therefore obviously does not work and hence should and could be OOP instead. 😉

Often, as I've mentioned other places, I like to slap a transaction on the entire request. This means that inproc stuff would run tin the same transaction anyway. Maybe the entities would have a nested one. (The UOW at leasst)  
The important bit is to know when to allow, if at all, mutating of entities being modified, and when to mandate that the handlers have to mind their own transactions and UOW instances. The latter will still allow for running stuff inproc.  

Hope my ramblings make sense.

`EntitySaving` -> `InProcHandlerMutatingEntityOrCanceling` -> Backoffice Domain / Content Domain / Data Layer  
`ContentPublished` -> `OutOfProcHandlerWritingNuCacheData` -> Front-End Domain / Cache Layer  

Dunno if @rustybox or @JimBobSquarePants agree it's an important distinction and have ideas on how to let the APIs communicate that?

Reply:
Thanks @lars-erik makes sense to me, I just wanted to know your standpoint. As mentioned above I agree with that. It's not how it is currently done with these extra events and we can refactor them out. 

> Events that mutate the entities could/should be ran in the same transaction.

Yes that is what we have today so that is fine.

> Events that perform operations in other bounded context should be possible to run OOP and in separate transactions. IE. write NuCache source data.

yep so we're all in agreement on this then

Reply:
@lars-erik what's your thoughts on out of proc events, do you see them being distributed via event aggregator or are you thinking more a seperate service bus + broker?

Reply:
Man, I should have not bothered building the aggregator if no one is going to use it!

Reply:
But of course we will use it for in proc events e.g ing (saving/publishing) events, my question is what’s the vision for the other events perhaps that’s still eventaggregator or perhaps there are separate classes to communicate between bounded contexts.

Reply:
The event aggregator is what I'm asking between the lines. Do we pass all events to it? Do we mark events in some way to indicate whether they're safe to mutate? Should the handlers be marked in some way? Is it relevant? Should there be one per bounded context / domain?  
As of now I think all events should go through the same aggregator and have the same abstractions. The parameter class either have references to mutable things or not. The latter might possibly also be immutable "by design" just to indicate that it's supposed to be serialized and handled out of proc.  
For entities the DbContext events should translate to something for use with the aggregator. IE. UmbracoEntitiesSaved with a copy of the change tracker's entity references. (I'm presuming we're going for EF?)  
IMO it's important to prevent any leakage of any library (MS or not) into the domain events.

Reply:
>[Edit]: Missed question from @rustybox above. I think everything should go through the event aggregator as mentioned. If you want to do something completely out of proc, then you make a handler to pass the message on to whatever you want to distribute via. A queue, a message bus, whatever.
This is all enabled by having a core aggregator (or more...) / command handling framework.
It's basically why we do this. 😁

YES!!!

All events should be triggered by the aggregator.
`INotificationHandler` instances then do whatever with the info.
Whether an event is mutable is determined by the properties of the passed `INotification` instance. If you don't want the properties to be mutable make them get only.

Reply:
All good. So we're saying "The parameter class either have references to mutable things or not."

Reply:
Yep, exactly. That's the power of it. You enforce the rules you want on a per instance basis.

Reply:
Coolio

Reply:
```csharp
public class EntitiesSaving : INotification
{
    public IEnumerable<IEntities> ChangedEntities { get; }
}

public class ContentPublished : INotification
{
    public IEnumerable<ContentDto> Content { get; }
}
```

Reply:
No more giant inheritance chain?
```csharp
ContentSavedEventArgs : SaveEventArgs<IContent>
SaveEventArgs<TEntity> : CancellableEnumerableObjectEventArgs<TEntity>
CancellableEnumerableObjectEventArgs<TEventObject> : CancellableObjectEventArgs<IEnumerable<TEventObject>>
CancellableObjectEventArgs<TEventObject> : CancellableObjectEventArgs
CancellableObjectEventArgs : CancellableEventArgs
CancellableEventArgs : EventArgs
```

Reply:
My favorite way of cancelling event pipelines:

`throw new Exception("This does not compute!")`